common/thorhelper/thorread.cpp

```diff
@@ -556,11 +703,458 @@ size32_t BinaryDiskRowReader::getFixedDiskRecordSize()
+        const RtlRecord &outRecord = mapping->queryProjectedMeta()->queryRecordAccessor(true);
+        translator.setown(createRecordTranslator(outRecord, inRecord));
+    }
+    IMPLEMENT_IINTERFACE_USING(CInterfaceOf<IAllocRowStream>)
+
+    virtual IRawRowStream * queryRawRowStream()
+    {
+        return this;
+    }
+
+    virtual IAllocRowStream * queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator)
+    {
+        allocatedBuilder.setAllocator(_outputAllocator);
+        outputAllocator.set(_outputAllocator);
+        return this;
+    }
+
+    virtual bool matches(const char * _format, bool _streamRemote, IDiskReadMapping * _mapping)
+    {
+        return false;
+    }
+
+    virtual void clearInput()
+    {
+        inputReader->clearInput();
+        rawInputStream = nullptr;
+    }
+
+    virtual bool setInputFile(const char * localFilename, const char * logicalFilename, unsigned partNumber, offset_t baseOffset, const IPropertyTree * meta, const FieldFilterArray & expectedFilter)
+    {
+        if (inputReader->setInputFile(localFilename, logicalFilename, partNumber, baseOffset, meta, expectedFilter))
+        {
+            rawInputStream = inputReader->queryRawRowStream();
+            return true;
+        }
+        return false;
+    }
+
+    virtual bool setInputFile(const RemoteFilename & filename, const char * logicalFilename, unsigned partNumber, offset_t baseOffset, const IPropertyTree * meta, const FieldFilterArray & expectedFilter)
+    {
+        if (inputReader->setInputFile(filename, logicalFilename, partNumber, baseOffset, meta, expectedFilter))
+        {
+            rawInputStream = inputReader->queryRawRowStream();
+            return true;
+        }
+        return false;
+    }
+
+//interface IRowReader
+    virtual bool getCursor(MemoryBuffer & cursor) { return rawInputStream->getCursor(cursor); }
+    virtual void setCursor(MemoryBuffer & cursor) { rawInputStream->setCursor(cursor); }
+    virtual void stop() { rawInputStream->stop(); }
+
+    virtual const void *nextRow(size32_t & resultSize) override
+    {
+        size32_t rawInputSize;
+        const void * next = rawInputStream->nextRow(rawInputSize);
+        if (isSpecialRow(next))
+            return next;
+
+        //MORE: optimize the case where fields are lost off the end, and not bother translating - but return the modified size.
+        tempOutputBuffer.clear();
+        resultSize = translator->translate(bufferBuilder, unexpectedCallback, (const byte *)next);
+        const void * ret = bufferBuilder.getSelf();
+        bufferBuilder.finishRow(resultSize);
+        return ret;
+    }
+
+    virtual const void *nextRow() override
+    {
+        size32_t rawInputSize;
+        const void * next = rawInputStream->nextRow(rawInputSize);
+        if (isSpecialRow(next))
+            return next;
+
+        size32_t size = translator->translate(allocatedBuilder.ensureRow(), unexpectedCallback, (const byte *)next);
```

Comment:
Should you release next?

Reply:
no, because it is calling the interface that returns a row that is only valid until the next call.